radio/src/pulses/afhds3.h

```diff
@@ -0,0 +1,523 @@
+  MIN_16dBm = 0x00,
+  BIND_POWER_FIRST = MIN_16dBm,
+  MIN_5dBm = 0x01,
+  MIN_0dbm = 0x02,
+  PLUS_5dBm = 0x03,
+  PLUS_14dBm = 0x04,
+  BIND_POWER_LAST = PLUS_14dBm,
+};
+
+enum RUN_POWER {
+  PLUS_15bBm = 0x00,
+  RUN_POWER_FIRST = PLUS_15bBm,
+  PLUS_20bBm = 0x01,
+  PLUS_27dbm = 0x02,
+  PLUS_30dBm = 0x03,
+  PLUS_33dBm = 0x04,
+  RUN_POWER_LAST = PLUS_33dBm,
+};
+
+enum EMI_STANDARD {
+  FCC = 0x00, CE = 0x01
+};
+
+enum TELEMETRY {
+  TELEMETRY_DISABLED = 0x00, TELEMETRY_ENABLED = 0x01
+};
+
+enum PULSE_MODE {
+  PWM_MODE = 0x00, PPM_MODE = 0x01,
+};
+
+enum SERIAL_MODE {
+  IBUS = 0x00, SBUS_MODE = 0x02
+};
+
+PACK(struct Config_s {
+  uint8_t bindPower;
+  uint8_t runPower;
+  uint8_t emiStandard;
+  uint8_t telemetry;
+  uint16_t pwmFreq;
+  uint8_t pulseMode;
+  uint8_t serialMode;
+  uint8_t channelCount;
+  uint16_t failSafeTimout;
+  int16_t failSafeMode[MAX_CHANNELS];
+});
+
+union Config_u {
+  Config_s config;
+  uint8_t buffer[sizeof(Config_s)];
+};
+
+enum CHANNELS_DATA_MODE {
+  CHANNELS = 0x01, FAIL_SAFE = 0x02,
+};
+
+PACK(struct ChannelsData {
+  uint8_t mode;
+  uint8_t channelsNumber;
+  int16_t data[MAX_CHANNELS];
+});
+
+union ChannelsData_u {
+  ChannelsData data;
+  uint8_t buffer[sizeof(ChannelsData)];
+};
+
+PACK(struct TelemetryData {
+  uint8_t sensorType;
+  uint8_t length;
+  uint8_t type;
+  uint8_t semsorID;
+  uint8_t data[8];
+});
+
+enum MODULE_POWER_SOURCE {
+  INTERNAL = 0x01, EXTERNAL = 0x02,
+};
+
+PACK(struct ModuleVersion {
+  uint32_t productNumber;
+  uint32_t hardwereVersion;
+  uint32_t bootloaderVersion;
+  uint32_t firmwareVersion;
+  uint32_t rfVersion;
+});
+
+PACK(struct CommandResult_s {
+  uint16_t command;
+  uint8_t result;
+  uint8_t respLen;
+});
+
+union AfhdsFrameData {
+  uint8_t value;
+  Config_s Config;
+  ChannelsData Channels;
+  TelemetryData Telemetry;
+  ModuleVersion Version;
+  CommandResult_s CommandResult;
+};
+
+PACK(struct AfhdsFrame {
+  uint8_t startByte;
+  uint8_t address;
+  uint8_t frameNumber;
+  uint8_t frameType;
+  uint8_t command;
+  uint8_t value;
+
+  AfhdsFrameData* GetData() {
+    return reinterpret_cast<AfhdsFrameData*>(&value);
+  }
+});
+
+#define FRM302_STATUS 0x56
+
+enum State {
+  UNKNOWN = 0,
+  SENDING_COMMAND,
+  AWAITING_RESPONSE,
+  IDLE
+};
+
+//one byte frames for request queue
+
+struct Frame {
+  enum COMMAND command;
+  enum FRAME_TYPE frameType;
+  uint8_t payload;
+  uint8_t frameNumber;
+  bool useFrameNumber;
+  uint8_t payloadSize;
+
+};
+
+
+//simple fifo implementation because Pulses is used as member of union and can not be non trivial type
+struct CommandFifo {
+  Frame commandFifo[8];
+  volatile uint32_t setIndex;
+  volatile uint32_t getIndex;
+
+  void clearCommandFifo();
+  inline uint32_t nextIndex(uint32_t idx)
+  {
+    return (idx + 1) & (sizeof(commandFifo)/sizeof(commandFifo[0]) - 1);
+  }
+  inline uint32_t prevIndex(uint32_t idx)
+  {
+    if(0) return (sizeof(commandFifo)/sizeof(commandFifo[0]) - 1);
```

Comment:
needed?

Reply:
Not used now - thus I have not spotted it. Fixed by checking idx against 0. Can be needed in feature.